include/zephyr/drivers/rtc.h

```diff
@@ -0,0 +1,527 @@
+		(const struct rtc_driver_api *)dev->api;
+
+	return api->alarm_is_pending(dev, id);
+}
+
+/**
+ * @brief API for setting alarm callback
+ *
+ * @param dev Device instance
+ * @param id Id of the alarm for which the callback shall be set
+ * @param callback Callback called when alarm occurs
+ * @param user_data Optional user data passed to callback
+ *
+ * @return 0 if successful
+ * @return -EINVAL if id is out of range
+ * @return -ENOTSUP if API is not supported by hardware
+ * @return -errno code if failure
+ */
+__syscall int rtc_alarm_set_callback(const struct device *dev, uint16_t id,
+				     rtc_alarm_callback callback, void *user_data);
+
+static inline int z_impl_rtc_alarm_set_callback(const struct device *dev, uint16_t id,
+						rtc_alarm_callback callback, void *user_data)
+{
+	const struct rtc_driver_api *api =
+		(const struct rtc_driver_api *)dev->api;
+
+	return api->alarm_set_callback(dev, id, callback, user_data);
+}
+
+/**
+ * @brief API for enabling RTC alarm callback
+ *
+ * @param dev Device instance
+ * @param id Id of the alarm for which the callback shall be enabled
+ *
+ * @return 0 if successful
+ * @return -EINVAL if id is out of range
+ * @return -ENOTSUP if API is not supported by hardware
+ * @return -errno code if failure
+ */
+__syscall int rtc_alarm_enable_callback(const struct device *dev, uint16_t id);
+
+static inline int z_impl_rtc_alarm_enable_callback(const struct device *dev, uint16_t id)
+{
+	const struct rtc_driver_api *api =
+		(const struct rtc_driver_api *)dev->api;
+
+	return api->alarm_enable_callback(dev, id);
+}
+
+/**
+ * @brief API for disabling RTC alarm callback
+ *
+ * @param dev Device instance
+ * @param id Id of the alarm for which the callback shall be enabled
+ *
+ * @return 0 if successful
+ * @return -EINVAL if id is out of range
+ * @return -ENOTSUP if API is not supported by hardware
+ * @return -errno code if failure
+ */
+__syscall int rtc_alarm_disable_callback(const struct device *dev, uint16_t id);
+
+static inline int z_impl_rtc_alarm_disable_callback(const struct device *dev, uint16_t id)
+{
+	const struct rtc_driver_api *api =
+		(const struct rtc_driver_api *)dev->api;
+
+	return api->alarm_disable_callback(dev, id);
+}
+
+/**
+ * @brief API for setting update callback
+ *
+ * @details The update occurs every time the RTC clock is updated by
+ * 1 second
+ *
+ * @param dev Device instance
+ * @param callback Callback called when update occurs
+ * @param user_data Optional user data passed to callback
+ *
+ * @return 0 if successful
+ * @return -ENOTSUP if API is not supported by hardware
+ * @return -errno code if failure
+ */
+__syscall int rtc_update_set_callback(const struct device *dev, rtc_update_callback callback,
+				      void *user_data);
+
+static inline int z_impl_rtc_update_set_callback(const struct device *dev,
+						 rtc_update_callback callback, void *user_data)
+{
+	const struct rtc_driver_api *api =
+		(const struct rtc_driver_api *)dev->api;
+
```

Comment:
As above - why must *every* RTC driver implement this, even if they do not support it, to just return an error?